// admin-auth/admin-utils.mjs
/**
 * Admin User Management Authorization Utilities
 * Shared functions for admin user CRUD operations and role-based access control
 */

// Valid role types for filtering
const VALID_ROLES = ["SA"];

/**
 * Check if a role is valid (SA, SPA_xxx, or TA_xxx)
 */
export const isValidRole = (role) =>
  VALID_ROLES.includes(role) || role.startsWith("SPA_") || role.startsWith("TA_");

/**
 * Extract and prioritize requester roles from event
 * Priority: SA > SPA_yyy > TA_xxx
 */
export const extractRolesFromEvent = (event) => {
  const jwtClaims = event.requestContext?.authorizer?.jwt?.claims;
  if (!jwtClaims) return [];

  let groups = jwtClaims["cognito:groups"];
  if (!groups) return [];

  // Handle string format (parse if needed)
  if (typeof groups === "string") {
    groups = groups.match(/[^\[\]\s]+/g);
  }

  const validRoles = Array.isArray(groups) ? groups.filter(isValidRole) : [];
  return prioritizeRoles(validRoles);
};

/**
 * Prioritize roles: SA > SPA_yyy > TA_xxx
 * Returns array with highest priority role(s)
 */
export const prioritizeRoles = (roles) => {
  if (!roles || roles.length === 0) return [];

  // SA takes priority
  if (roles.includes("SA")) {
    return ["SA"];
  }

  // SPA_yyy takes priority over TA_xxx
  const spaRole = roles.find(role => role.startsWith("SPA_"));
  if (spaRole) {
    return [spaRole];
  }

  // Return all TA_xxx roles
  const taRoles = roles.filter(role => role.startsWith("TA_"));
  return taRoles.length > 0 ? taRoles : [];
};

/**
 * Get user's groups from Cognito
 */
export const getUserGroups = async (username, cognitoISP) => {
  try {
    const { AdminListGroupsForUserCommand } = await import("@aws-sdk/client-cognito-identity-provider");
    
    const result = await cognitoISP.send(
      new AdminListGroupsForUserCommand({
        UserPoolId: process.env.USERPOOL_ID,
        Username: username,
        Limit: 60,
      }),
    );

    return (
      result.Groups?.map((group) => group.GroupName).filter(isValidRole) || []
    );
  } catch (error) {
    console.log("Error getting user groups:", error);
    return [];
  }
};

/**
 * Validate admin operation permission (GET, PUT, DELETE)
 * Returns { isValid: boolean, error?: string }
 */
export const validateAdminOperation = async (
  requesterRoles,
  operation,
  targetUsername,
  cognitoISP,
) => {
  console.log("Validating admin operation permission:", {
    requesterRoles,
    operation,
    targetUsername,
  });

  if (!requesterRoles?.length) {
    return { isValid: false, error: "No requester roles provided" };
  }

  if (!targetUsername) {
    return { isValid: false, error: "No target username provided" };
  }

  // Early return for SA - can do anything
  if (requesterRoles.includes("SA")) {
    return { isValid: true };
  }

  // Get target user's groups
  const targetUserGroups = await getUserGroups(targetUsername, cognitoISP);
  console.log("Target user groups:", targetUserGroups);

  const targetHasSA = targetUserGroups.includes("SA");
  const hasSPA = requesterRoles.some(role => role.startsWith("SPA_"));
  const spaRole = requesterRoles.find(role => role.startsWith("SPA_"));
  const requesterTARoles = requesterRoles.filter((role) =>
    role.startsWith("TA_"),
  );

  // SPA_yyy validation
  if (hasSPA) {
    return targetHasSA
      ? {
          isValid: false,
          error: `${spaRole} cannot perform ${operation} operation on SA users`,
        }
      : { isValid: true };
  }

  // TA validation
  if (requesterTARoles.length > 0) {
    const hasCommonGroup = targetUserGroups.some((group) =>
      requesterTARoles.includes(group),
    );
    return hasCommonGroup
      ? { isValid: true }
      : {
          isValid: false,
          error: `TA users with roles [${requesterTARoles.join(", ")}] can only perform operations on users in their groups. Target user has groups: [${targetUserGroups.join(", ")}]`,
        };
  }

  // No valid roles found
  return {
    isValid: false,
    error: `Requester roles [${requesterRoles.join(", ")}] are not authorized to perform admin operations`,
  };
};

/**
 * Get available TA groups from Cognito
 */
export const getAvailableTAGroups = async (cognitoISP) => {
  try {
    const { ListGroupsCommand } = await import("@aws-sdk/client-cognito-identity-provider");
    
    const data = await cognitoISP.send(
      new ListGroupsCommand({ UserPoolId: process.env.USERPOOL_ID }),
    );
    
    if (data.Groups && data.Groups.length > 0) {
      return data.Groups
        .filter((group) => group.GroupName.startsWith("TA_"))
        .map((group) => group.GroupName);
    }
    return [];
  } catch (error) {
    console.log("Error getting TA groups:", error);
    return [];
  }
};

/**
 * Get available SPA groups from Cognito
 */
export const getAvailableSPAGroups = async (cognitoISP) => {
  try {
    const { ListGroupsCommand } = await import("@aws-sdk/client-cognito-identity-provider");
    
    const data = await cognitoISP.send(
      new ListGroupsCommand({ UserPoolId: process.env.USERPOOL_ID }),
    );
    
    if (data.Groups && data.Groups.length > 0) {
      return data.Groups
        .filter((group) => group.GroupName.startsWith("SPA_"))
        .map((group) => group.GroupName);
    }
    return [];
  } catch (error) {
    console.log("Error getting SPA groups:", error);
    return [];
  }
};

/**
 * Validate permission to create user with specific groups
 * Returns { isValid: boolean, error?: string }
 */
export const validateGroupCreationPermission = async (
  requesterRoles,
  targetGroups,
  cognitoISP,
) => {
  console.log("Validating group creation permission:", {
    requesterRoles,
    targetGroups,
  });

  if (!requesterRoles || requesterRoles.length === 0) {
    return { isValid: false, error: "No requester roles provided" };
  }

  if (!targetGroups || targetGroups.length === 0) {
    return { isValid: false, error: "No target groups specified" };
  }

  const availableTAGroups = await getAvailableTAGroups(cognitoISP);
  const availableSPAGroups = await getAvailableSPAGroups(cognitoISP);

  // Check if requester has SA or SPA_yyy role
  const hasSA = requesterRoles.includes("SA");
  const hasSPA = requesterRoles.some(role => role.startsWith("SPA_"));
  const spaRole = requesterRoles.find(role => role.startsWith("SPA_"));

  // Get all TA_xxx roles that the requester has
  const requesterTARoles = requesterRoles.filter(
    (role) => role.startsWith("TA_") && availableTAGroups.includes(role),
  );

  // Cache for SPA org-scoped TA groups (lazy-loaded once per call)
  let spaOrgTAGroups = null;

  for (const targetGroup of targetGroups) {
    let isAuthorized = false;
    let errorMessage = "";

    if (hasSA) {
      // SA can create users with any SPA_yyy or any TA_XXX group
      if (targetGroup.startsWith("SPA_") || availableTAGroups.includes(targetGroup)) {
        isAuthorized = true;
      } else {
        errorMessage = `SA cannot create user with group: ${targetGroup}. Allowed groups: ${availableSPAGroups.join(", ")}, ${availableTAGroups.join(", ")}`;
      }
    } else if (hasSPA) {
      // SPA_yyy can create users with their own SPA_yyy or TA groups for tenants in their org
      if (targetGroup === spaRole) {
        isAuthorized = true;
      } else if (targetGroup.startsWith("TA_")) {
        // Lazy-load org-scoped TA groups (only once)
        if (!spaOrgTAGroups) {
          const orgId = spaRole.substring(4);
          spaOrgTAGroups = await getTAGroupsForOrg(orgId, availableTAGroups);
        }
        if (spaOrgTAGroups.includes(targetGroup)) {
          isAuthorized = true;
        } else {
          errorMessage = `${spaRole} cannot create user with group: ${targetGroup}. This tenant does not belong to your org. Allowed TA groups: ${spaOrgTAGroups.join(", ")}`;
        }
      } else {
        errorMessage = `${spaRole} cannot create user with group: ${targetGroup}. Allowed groups: ${spaRole}, org-scoped TA groups`;
      }
    } else if (requesterTARoles.length > 0) {
      // TA_XXX users can create users with any of their TA_XXX groups
      if (requesterTARoles.includes(targetGroup)) {
        isAuthorized = true;
      } else {
        errorMessage = `Requester with roles [${requesterTARoles.join(", ")}] can only create users with groups: ${requesterTARoles.join(", ")}. Cannot create user with group: ${targetGroup}`;
      }
    } else {
      // No valid roles found
      errorMessage = `Requester roles [${requesterRoles.join(", ")}] are not authorized to create users`;
    }

    if (!isAuthorized) {
      return { isValid: false, error: errorMessage };
    }
  }

  return { isValid: true };
};

/**
 * Get TA groups scoped to a specific org (tenants belonging to that org)
 * @param {string} orgId - Organization ID
 * @param {string[]} allTAGroups - All available TA groups from Cognito
 * @returns {Promise<string[]>} TA groups for tenants in the org
 */
export const getTAGroupsForOrg = async (orgId, allTAGroups) => {
  try {
    const { DynamoDBClient, QueryCommand } = await import('@aws-sdk/client-dynamodb');
    const tableName = process.env.AMFATENANT_TABLE || 'amfa-tenanttable';
    const ddb = new DynamoDBClient({ region: process.env.AWS_REGION });
    const result = await ddb.send(new QueryCommand({
      TableName: tableName,
      IndexName: 'org-id-index',
      KeyConditionExpression: 'org_id = :orgId',
      ExpressionAttributeValues: { ':orgId': { S: orgId } },
    }));
    const orgTenants = (result.Items || []).map(item => ({
      tenantId: item.id?.S,
      orgId: item.org_id?.S,
    }));
    const orgTenantIds = new Set(
      orgTenants.map(t => {
        // getTenantsByOrg returns tenantId which may have TENANT# prefix
        const id = t.tenantId || '';
        return id.replace('TENANT#', '');
      })
    );
    console.log(`[Scope] Org '${orgId}' tenants:`, [...orgTenantIds]);
    return allTAGroups.filter(group => {
      const tenantId = group.substring(3); // Remove 'TA_' prefix
      return orgTenantIds.has(tenantId);
    });
  } catch (error) {
    console.error(`[Scope] Error getting TA groups for org '${orgId}':`, error.message);
    return [];
  }
};

/**
 * Get available groups that a user with given roles can assign
 * Used by frontend to populate group selection dropdowns
 */
export const getAvailableTAGroupsForRole = async (
  requesterRoles,
  cognitoISP,
) => {
  if (!requesterRoles || requesterRoles.length === 0) {
    return { groups: [], error: "No requester roles provided" };
  }

  const availableTAGroups = await getAvailableTAGroups(cognitoISP);
  const availableSPAGroups = await getAvailableSPAGroups(cognitoISP);

  // Check if requester has SA or SPA_yyy role
  const hasSA = requesterRoles.includes("SA");
  const hasSPA = requesterRoles.some(role => role.startsWith("SPA_"));
  const spaRole = requesterRoles.find(role => role.startsWith("SPA_"));

  // Get all TA_xxx roles that the requester has
  const requesterTARoles = requesterRoles.filter(
    (role) => role.startsWith("TA_") && availableTAGroups.includes(role),
  );

  if (hasSA) {
    // SA can create users with any SPA_yyy or any TA_XXX group
    return { groups: [...availableSPAGroups, ...availableTAGroups] };
  } else if (hasSPA) {
    // SPA_yyy can only assign their own SPA_yyy or TA groups for tenants in their org
    const orgId = spaRole.substring(4); // Extract orgId from SPA_<orgId>
    const orgTAGroups = await getTAGroupsForOrg(orgId, availableTAGroups);
    console.log(`[Scope] SPA '${spaRole}' available groups: [${spaRole}, ${orgTAGroups.join(', ')}]`);
    return { groups: [spaRole, ...orgTAGroups] };
  } else if (requesterTARoles.length > 0) {
    // TA_XXX users can create users with any of their TA_XXX groups
    // Remove duplicates and sort for consistency
    const uniqueGroups = [...new Set(requesterTARoles)].sort();
    return { groups: uniqueGroups };
  } else {
    return {
      groups: [],
      error: `Requester roles [${requesterRoles.join(", ")}] cannot create users`,
    };
  }
};

/**
 * Create standardized HTTP response
 */
export const createAdminResponse = (statusCode, data = null, error = null) => {
  const headers = {
    "Access-Control-Allow-Headers":
      "Content-Type,Authorization,X-Api-Key,X-Requested-With",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "OPTIONS,GET,POST,PUT,DELETE",
  };

  const body = error
    ? JSON.stringify({ type: error.type || "error", message: error.message })
    : JSON.stringify({ data });

  return {
    statusCode,
    headers,
    body,
  };
};
