// admin-auth/asm-utils.mjs
/**
 * ASM Portal integration utilities for deregistering admins.
 *
 * Provides shared helper functions to call ASM Portal APIs when
 * removing TA (Tenant Admin) or SPA (Service Provider Admin) roles.
 */

import {
  SecretsManagerClient,
  GetSecretValueCommand,
} from "@aws-sdk/client-secrets-manager";

import { DynamoDBClient, QueryCommand } from "@aws-sdk/client-dynamodb";

const secretsManager = new SecretsManagerClient({ region: process.env.AWS_REGION });
const dynamodb = new DynamoDBClient({ region: process.env.AWS_REGION });

/**
 * Deregister a Tenant Admin from ASM Portal by calling tenantAdmin.ap with action=remove.
 *
 * Reads the tenant's asmClientId and org's asmSecretKey from Secrets Manager,
 * then calls the ASM Portal tenantAdmin.ap endpoint.
 *
 * @param {string} email - The TA admin's email to remove
 * @param {string} tenantId - The tenant ID (internal ID, e.g., "mytenant")
 * @param {string} requestedBy - The email of the user who initiated the request
 * @returns {Promise<Object|null>} ASM API response or null if skipped
 */
export async function deregisterTAAdminFromASM(email, tenantId, requestedBy) {
  const asmPortalUrl = process.env.ASM_PORTAL_URL;
  if (!asmPortalUrl) {
    console.warn("[ASM] ASM_PORTAL_URL not configured, skipping TA admin deregistration from ASM");
    return null;
  }

  // Step 1: Read tenant ASM credentials from Secrets Manager to get asmClientId and orgId
  const tenantSecretName = `apersona/asm/tenant/${tenantId}`;
  console.log(`[ASM] Reading tenant credentials from ${tenantSecretName}...`);

  let tenantCredentials;
  try {
    const secretResult = await secretsManager.send(
      new GetSecretValueCommand({ SecretId: tenantSecretName }),
    );
    tenantCredentials = JSON.parse(secretResult.SecretString);
  } catch (secretError) {
    console.error(`[ASM] Failed to read tenant credentials (${tenantSecretName}): ${secretError.message}`);
    return null;
  }

  const asmClientId = tenantCredentials.asmClientId;
  const orgId = tenantCredentials.orgId;
  if (!asmClientId) {
    console.warn(`[ASM] No asmClientId found in tenant credentials for tenant '${tenantId}', skipping`);
    return null;
  }
  if (!orgId) {
    console.warn(`[ASM] No orgId found in tenant credentials for tenant '${tenantId}', skipping`);
    return null;
  }

  // Step 2: Read org ASM credentials from Secrets Manager to get asmSecretKey
  const orgSecretName = `apersona/asm/org/${orgId}`;
  console.log(`[ASM] Reading org credentials from ${orgSecretName}...`);

  let orgCredentials;
  try {
    const secretResult = await secretsManager.send(
      new GetSecretValueCommand({ SecretId: orgSecretName }),
    );
    orgCredentials = JSON.parse(secretResult.SecretString);
  } catch (secretError) {
    console.error(`[ASM] Failed to read org credentials (${orgSecretName}): ${secretError.message}`);
    return null;
  }

  const asmSecretKey = orgCredentials.asmSecretKey;
  if (!asmSecretKey) {
    console.warn(`[ASM] No asmSecretKey found in org credentials for org '${orgId}', skipping`);
    return null;
  }

  // Step 3: Read tenant name from DynamoDB
  let tenantName = tenantId; // fallback to tenantId if lookup fails
  try {
    const queryResult = await dynamodb.send(new QueryCommand({
      TableName: "amfa-tenanttable",
      KeyConditionExpression: "id = :id AND begins_with(sk, :sk_prefix)",
      ExpressionAttributeValues: {
        ":id": { S: `TENANT#${tenantId}` },
        ":sk_prefix": { S: "TENANT#" },
      },
    }));
    if (queryResult.Items && queryResult.Items.length > 0) {
      tenantName = queryResult.Items[0].name?.S || tenantId;
    }
  } catch (dbError) {
    console.warn(`[ASM] Could not read tenant name from DynamoDB (non-fatal): ${dbError.message}`);
  }

  const awsAccountId = process.env.ACCOUNT_ID || "";

  console.log(`[ASM] Deregistering TA admin from ASM portal...`);
  console.log(`[ASM]   URL: ${asmPortalUrl}/tenantAdmin.ap`);
  console.log(`[ASM]   tenantId (asmClientId): ${asmClientId}`);
  console.log(`[ASM]   tenantName: ${tenantName}`);
  console.log(`[ASM]   tenantAdminEmail: ${email}`);
  console.log(`[ASM]   action: remove`);
  console.log(`[ASM]   requestedBy: ${requestedBy}`);

  const formData = new URLSearchParams({
    tenantId: asmClientId,
    tenantName: tenantName,
    tenantAdminEmail: email,
    action: "remove",
    requestedBy: requestedBy,
    awsAccountId: awsAccountId,
    asmSecretKey: asmSecretKey,
  });

  const response = await fetch(`${asmPortalUrl}/tenantAdmin.ap`, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: formData.toString(),
  });

  if (!response.ok) {
    const errorText = await response.text();
    console.error(`[ASM] tenantAdmin.ap (action=remove) failed (${response.status}): ${errorText}`);
    return null;
  }

  const result = await response.json();
  console.log(`[ASM] ✓ TA admin deregistered from ASM:`, JSON.stringify(result));

  return result;
}

/**
 * Deregister a Service Provider Admin from ASM Portal by calling removeServiceProviderAdmin.ap.
 *
 * Reads the org's serviceProviderId and asmSecretKey from Secrets Manager,
 * then calls the ASM Portal removeServiceProviderAdmin.ap endpoint.
 *
 * @param {string} email - The SPA admin's email to remove
 * @param {string} orgId - The organization ID (from SPA_<orgId> group)
 * @param {string} requestedBy - The email of the user who initiated the request
 * @returns {Promise<Object|null>} ASM API response or null if skipped
 */
export async function deregisterSPAAdminFromASM(email, orgId, requestedBy) {
  const asmPortalUrl = process.env.ASM_PORTAL_URL;
  if (!asmPortalUrl) {
    console.warn("[ASM] ASM_PORTAL_URL not configured, skipping SPA admin deregistration from ASM");
    return null;
  }

  // Read org ASM credentials from Secrets Manager
  const orgSecretName = `apersona/asm/org/${orgId}`;
  console.log(`[ASM] Reading org credentials from ${orgSecretName}...`);

  let orgCredentials;
  try {
    const secretResult = await secretsManager.send(
      new GetSecretValueCommand({ SecretId: orgSecretName }),
    );
    orgCredentials = JSON.parse(secretResult.SecretString);
  } catch (secretError) {
    console.error(`[ASM] Failed to read org credentials (${orgSecretName}): ${secretError.message}`);
    return null;
  }

  const serviceProviderId = orgCredentials.serviceProviderId;
  const asmSecretKey = orgCredentials.asmSecretKey;

  if (!serviceProviderId) {
    console.warn(`[ASM] No serviceProviderId found in org credentials for org '${orgId}', skipping`);
    return null;
  }
  if (!asmSecretKey) {
    console.warn(`[ASM] No asmSecretKey found in org credentials for org '${orgId}', skipping`);
    return null;
  }

  const awsAccountId = process.env.ACCOUNT_ID || "";

  console.log(`[ASM] Deregistering SPA admin from ASM portal...`);
  console.log(`[ASM]   URL: ${asmPortalUrl}/removeServiceProviderAdmin.ap`);
  console.log(`[ASM]   email: ${email}`);
  console.log(`[ASM]   serviceProviderId: ${serviceProviderId}`);
  console.log(`[ASM]   requestedBy: ${requestedBy}`);

  const formData = new URLSearchParams({
    email: email,
    serviceProviderId: serviceProviderId,
    awsAccountId: awsAccountId,
    awsSecretKey: asmSecretKey,
    requestedBy: requestedBy,
  });

  const response = await fetch(`${asmPortalUrl}/removeServiceProviderAdmin.ap`, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: formData.toString(),
  });

  if (!response.ok) {
    const errorText = await response.text();
    console.error(`[ASM] removeServiceProviderAdmin.ap failed (${response.status}): ${errorText}`);
    return null;
  }

  const result = await response.json();
  console.log(`[ASM] ✓ SPA admin deregistered from ASM:`, JSON.stringify(result));

  return result;
}
