// admin-auth/authorize.mjs
import { DynamoDBClient, QueryCommand } from '@aws-sdk/client-dynamodb';

const dynamodb = new DynamoDBClient({ region: process.env.AWS_REGION });

/**
 * Validate tenant access based on user role
 * @param {Object} event - API Gateway event
 * @param {string} tenantId - Tenant ID to validate access for
 * @returns {Promise<Object>} { authorized: boolean, userPoolId: string, tenantOrgId: string, error: string }
 */
export async function validateTenantAccess(event, tenantId) {
    // Extract role from JWT (cognito:groups)
    // Support both HttpUserPoolAuthorizer (jwt.claims) and custom authorizer formats
    let groups = event.requestContext?.authorizer?.jwt?.claims?.['cognito:groups'] 
        || event.requestContext?.authorizer?.['cognito:groups'] 
        || [];
    
    // Handle string format from JWT claims (e.g., "[SA]" or "[SPA_org1]")
    if (typeof groups === 'string') {
        groups = groups.replace(/^\[|\]$/g, '').split(',').map(g => g.trim()).filter(Boolean);
    }
    
    const role = Array.isArray(groups) ? groups[0] : groups; // SA, SPA_yyy, or TA_xxx
    
    if (!role) {
        return { 
            authorized: false, 
            error: 'No role found in JWT',
            statusCode: 401 
        };
    }
    
    // Query tenant info from DynamoDB using composite key
    const tenantResult = await dynamodb.send(new QueryCommand({
        TableName: process.env.AMFATENANT_TABLE,
        KeyConditionExpression: 'id = :id AND begins_with(sk, :sk_prefix)',
        ExpressionAttributeValues: {
            ':id': { S: `TENANT#${tenantId}` },
            ':sk_prefix': { S: 'TENANT#' }
        }
    }));
    
    const tenantItem = tenantResult.Items && tenantResult.Items.length > 0 ? tenantResult.Items[0] : null;
    
    if (!tenantItem) {
        return { 
            authorized: false, 
            error: `Tenant ${tenantId} not found`,
            statusCode: 404 
        };
    }
    
    const tenantOrgId = tenantItem.org_id?.S;
    const userPoolId = tenantItem.userpool?.S;
    
    // Authorization logic based on role
    if (role === 'SA') {
        // Super Admin - full access to all tenants
        return { 
            authorized: true, 
            userPoolId, 
            tenantOrgId,
            role,
            statusCode: 200 
        };
        
    } else if (role?.startsWith('SPA_')) {
        // Service Provider Admin - can access tenants in their org
        const orgId = role.substring(4); // Extract 'yyy' from 'SPA_yyy'
        
        if (!tenantOrgId) {
            return {
                authorized: false,
                error: `Tenant ${tenantId} has no org_id assigned`,
                statusCode: 403
            };
        }
        
        if (tenantOrgId === orgId) {
            return { 
                authorized: true, 
                userPoolId, 
                tenantOrgId,
                role,
                statusCode: 200 
            };
        }
        
        return { 
            authorized: false, 
            error: `SPA_${orgId} cannot access tenant ${tenantId} (belongs to org ${tenantOrgId})`,
            statusCode: 403 
        };
        
    } else if (role?.startsWith('TA_')) {
        // Tenant Admin - can ONLY access their own tenant
        const allowedTenant = role.substring(3); // Extract 'xxx' from 'TA_xxx'
        
        if (tenantId === allowedTenant) {
            return { 
                authorized: true, 
                userPoolId, 
                tenantOrgId,
                role,
                statusCode: 200 
            };
        }
        
        return { 
            authorized: false, 
            error: `TA_${allowedTenant} cannot access tenant ${tenantId}`,
            statusCode: 403 
        };
        
    } else {
        // Unknown/invalid role
        return { 
            authorized: false, 
            error: `Invalid role: ${role}`,
            statusCode: 403 
        };
    }
}

/**
 * Extract tenant_id from API request
 * Checks query parameters, path parameters, and request body
 * @param {Object} event - API Gateway event
 * @returns {string|null} tenant_id or null if not found
 */
export function getTenantIdFromRequest(event) {
    // Try X-Tenant-Id header first (primary method from frontend)
    const headers = event.headers || {};
    const tenantIdHeader = headers['x-tenant-id'] || headers['X-Tenant-Id'];
    if (tenantIdHeader) {
        return tenantIdHeader;
    }
    
    // Try query parameters
    if (event.queryStringParameters?.tenant_id) {
        return event.queryStringParameters.tenant_id;
    }
    
    // Try path parameters
    if (event.pathParameters?.tenant_id) {
        return event.pathParameters.tenant_id;
    }
    
    // Try request body
    try {
        const body = JSON.parse(event.body || '{}');
        if (body.tenant_id) {
            return body.tenant_id;
        }
    } catch (error) {
        // Invalid JSON body, continue
    }
    
    return null;
}
