// admin-auth/index.cjs
// CJS wrapper for ESM admin-auth module
// This allows CommonJS bundled lambdas (like samlslist/dist) to use this layer

let _module;

async function loadModule() {
  if (!_module) {
    _module = await import('./index.mjs');
  }
  return _module;
}

// Re-export all functions as async wrappers
module.exports.validateTenantAccess = async function(...args) {
  const mod = await loadModule();
  return mod.validateTenantAccess(...args);
};

module.exports.getTenantIdFromRequest = function(...args) {
  // This is synchronous in the ESM version, so we load module eagerly
  // For CJS consumers, we provide a sync fallback that reads headers directly
  const event = args[0];
  if (!event) return null;
  
  // Try X-Tenant-Id header first (primary method from frontend)
  const headers = event.headers || {};
  const tenantIdHeader = headers['x-tenant-id'] || headers['X-Tenant-Id'];
  if (tenantIdHeader) return tenantIdHeader;
  
  // Try query parameters
  if (event.queryStringParameters?.tenant_id) return event.queryStringParameters.tenant_id;
  
  // Try path parameters
  if (event.pathParameters?.tenant_id) return event.pathParameters.tenant_id;
  
  // Try request body
  try {
    const body = JSON.parse(event.body || '{}');
    if (body.tenant_id) return body.tenant_id;
  } catch (error) {
    // Invalid JSON body, continue
  }
  
  return null;
};

module.exports.createResponse = function(statusCode, body, tenantId) {
  return {
    statusCode,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': 'Content-Type,Authorization,X-Api-Key,X-Tenant-Id,Content-Range,X-Requested-With',
      'Access-Control-Allow-Methods': 'OPTIONS,GET,POST,PUT,DELETE',
      'Access-Control-Expose-Headers': 'Content-Range',
    },
    body: JSON.stringify(body),
  };
};

module.exports.getTenantInfo = async function(...args) {
  const mod = await loadModule();
  return mod.getTenantInfo(...args);
};

module.exports.getTenantsByOrg = async function(...args) {
  const mod = await loadModule();
  return mod.getTenantsByOrg(...args);
};

module.exports.isValidRole = function(...args) {
  // Sync function - inline implementation
  const role = args[0];
  if (!role) return false;
  return role === 'SA' || role.startsWith('SPA_') || role.startsWith('TA_');
};

module.exports.extractRolesFromEvent = function(event) {
  // Sync implementation matching admin-utils.mjs
  const jwtClaims = event?.requestContext?.authorizer?.jwt?.claims;
  if (!jwtClaims) return [];

  let groups = jwtClaims["cognito:groups"];
  if (!groups) return [];

  // Handle string format (parse if needed)
  if (typeof groups === "string") {
    groups = groups.match(/[^\[\]\s]+/g);
  }

  const validRoles = Array.isArray(groups)
    ? groups.filter(role => role === 'SA' || role.startsWith('SPA_') || role.startsWith('TA_'))
    : [];
  return module.exports.prioritizeRoles(validRoles);
};

module.exports.prioritizeRoles = function(roles) {
  if (!roles || roles.length === 0) return [];
  if (roles.includes("SA")) return ["SA"];
  const spaRole = roles.find(role => role.startsWith("SPA_"));
  if (spaRole) return [spaRole];
  const taRoles = roles.filter(role => role.startsWith("TA_"));
  return taRoles.length > 0 ? taRoles : [];
};

module.exports.getUserGroups = async function(...args) {
  const mod = await loadModule();
  return mod.getUserGroups(...args);
};

module.exports.validateAdminOperation = async function(...args) {
  const mod = await loadModule();
  return mod.validateAdminOperation(...args);
};

module.exports.getAvailableTAGroups = async function(...args) {
  const mod = await loadModule();
  return mod.getAvailableTAGroups(...args);
};

module.exports.getAvailableSPAGroups = async function(...args) {
  const mod = await loadModule();
  return mod.getAvailableSPAGroups(...args);
};

module.exports.validateGroupCreationPermission = async function(...args) {
  const mod = await loadModule();
  return mod.validateGroupCreationPermission(...args);
};

module.exports.getAvailableTAGroupsForRole = async function(...args) {
  const mod = await loadModule();
  return mod.getAvailableTAGroupsForRole(...args);
};

module.exports.getTAGroupsForOrg = async function(...args) {
  const mod = await loadModule();
  return mod.getTAGroupsForOrg(...args);
};

module.exports.createAdminResponse = function(statusCode, data, error) {
  // Sync implementation matching admin-utils.mjs
  const headers = {
    "Access-Control-Allow-Headers": "Content-Type,Authorization,X-Api-Key,X-Requested-With",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "OPTIONS,GET,POST,PUT,DELETE",
  };
  const body = error
    ? JSON.stringify({ type: error.type || "error", message: error.message })
    : JSON.stringify({ data });
  return { statusCode, headers, body };
};

// Structured logger for JSON CloudWatch Logs (2.2)
// CJS version matching logger.mjs API: createLogger(serviceName, defaultKeys)
const _LOG_LEVELS = { DEBUG: 0, INFO: 1, WARN: 2, ERROR: 3 };

module.exports.createLogger = function(serviceName, defaultKeys = {}) {
  const configuredLevel = (process.env.LOG_LEVEL || 'INFO').toUpperCase();
  const minLevel = _LOG_LEVELS[configuredLevel] ?? _LOG_LEVELS.INFO;
  let persistentKeys = { ...defaultKeys };

  function log(level, message, extra = {}) {
    if ((_LOG_LEVELS[level] ?? 0) < minLevel) return;
    const entry = {
      timestamp: new Date().toISOString(),
      level,
      service: serviceName,
      message,
      ...persistentKeys,
      ...extra,
    };
    Object.keys(entry).forEach(k => { if (entry[k] === undefined) delete entry[k]; });
    console.log(JSON.stringify(entry));
  }

  return {
    info: (message, extra) => log('INFO', message, extra),
    warn: (message, extra) => log('WARN', message, extra),
    error: (message, extra) => log('ERROR', message, extra),
    debug: (message, extra) => log('DEBUG', message, extra),
    appendKeys: (keys) => { persistentKeys = { ...persistentKeys, ...keys }; },
    resetKeys: () => { persistentKeys = { ...defaultKeys }; },
  };
};
