// admin-auth/logger.mjs
/**
 * Structured Logger for AMFA Admin Portal Lambdas (2.2)
 *
 * Outputs JSON-formatted logs that are fully queryable via CloudWatch Logs Insights.
 * Zero dependencies — uses console.log with JSON.stringify.
 *
 * Usage:
 *   import { createLogger } from 'admin-auth';
 *   const logger = createLogger('my-lambda');
 *   logger.info('Operation completed', { tenantId: 'abc', duration: 1234 });
 *
 * Output:
 *   {"timestamp":"2026-04-04T10:00:00.000Z","level":"INFO","service":"my-lambda","message":"Operation completed","tenantId":"abc","duration":1234}
 *
 * CloudWatch Logs Insights queries:
 *   fields @timestamp, level, message, tenantId
 *   | filter service = "my-lambda" and level = "ERROR"
 *   | filter tenantId = "abc"
 *   | stats count(*) by tenantId
 */

const LOG_LEVELS = { DEBUG: 0, INFO: 1, WARN: 2, ERROR: 3 };

/**
 * Create a structured logger instance for a Lambda function.
 *
 * @param {string} serviceName - Lambda/service name (e.g., 'provision-tenant', 'userslist')
 * @param {Object} [defaultKeys={}] - Default keys to include in every log entry (e.g., { tenantId })
 * @returns {Object} Logger with info/warn/error/debug methods
 */
export function createLogger(serviceName, defaultKeys = {}) {
  const configuredLevel = (process.env.LOG_LEVEL || 'INFO').toUpperCase();
  const minLevel = LOG_LEVELS[configuredLevel] ?? LOG_LEVELS.INFO;
  let persistentKeys = { ...defaultKeys };

  function log(level, message, extra = {}) {
    if ((LOG_LEVELS[level] ?? 0) < minLevel) return;

    const entry = {
      timestamp: new Date().toISOString(),
      level,
      service: serviceName,
      message,
      ...persistentKeys,
      ...extra,
    };

    // Remove undefined values to keep logs clean
    Object.keys(entry).forEach(k => {
      if (entry[k] === undefined) delete entry[k];
    });

    console.log(JSON.stringify(entry));
  }

  return {
    /** Log at INFO level */
    info: (message, extra) => log('INFO', message, extra),
    /** Log at WARN level */
    warn: (message, extra) => log('WARN', message, extra),
    /** Log at ERROR level */
    error: (message, extra) => log('ERROR', message, extra),
    /** Log at DEBUG level (only when LOG_LEVEL=DEBUG) */
    debug: (message, extra) => log('DEBUG', message, extra),
    /**
     * Add persistent keys to all subsequent log entries.
     * Useful for adding tenantId, requestId, etc. at handler start.
     * @param {Object} keys - Key-value pairs to persist
     */
    appendKeys: (keys) => {
      persistentKeys = { ...persistentKeys, ...keys };
    },
    /**
     * Reset persistent keys (e.g., between invocations in warm Lambda)
     */
    resetKeys: () => {
      persistentKeys = { ...defaultKeys };
    },
  };
}
