// admin-auth/tenant-utils.mjs
import { DynamoDBClient, QueryCommand } from '@aws-sdk/client-dynamodb';
import { getCorsHeaders } from './cors-utils.mjs';

const dynamodb = new DynamoDBClient({ region: process.env.AWS_REGION });

/**
 * Get tenant information by tenant ID
 * @param {string} tenantId - Tenant ID
 * @returns {Promise<Object>} Tenant information
 */
export async function getTenantInfo(tenantId) {
    const result = await dynamodb.send(new QueryCommand({
        TableName: process.env.AMFATENANT_TABLE,
        KeyConditionExpression: 'id = :id AND begins_with(sk, :sk_prefix)',
        ExpressionAttributeValues: {
            ':id': { S: `TENANT#${tenantId}` },
            ':sk_prefix': { S: 'TENANT#' }
        }
    }));
    
    const item = result.Items && result.Items.length > 0 ? result.Items[0] : null;
    
    if (!item) {
        return null;
    }
    
    // Remove TENANT# prefix from id for return value
    const cleanTenantId = item.id?.S?.replace('TENANT#', '') || tenantId;
    
    return {
        tenantId: cleanTenantId,
        name: item.name?.S,
        orgId: item.org_id?.S,
        userPoolId: item.userpool?.S,
        endUserClientId: item.enduserclientid?.S,
        url: item.url?.S,
        endUserSpUrl: item.endUserSpUrl?.S,
        samlproxy: item.samlproxy?.BOOL,
        contact: item.contact?.S,
    };
}

/**
 * Get all tenants for a specific organization
 * @param {string} orgId - Organization ID
 * @returns {Promise<Array>} List of tenants in the organization
 */
export async function getTenantsByOrg(orgId) {
    const result = await dynamodb.send(new QueryCommand({
        TableName: process.env.AMFATENANT_TABLE,
        IndexName: 'org-id-index',
        KeyConditionExpression: 'org_id = :orgId',
        ExpressionAttributeValues: {
            ':orgId': { S: orgId }
        }
    }));
    
    if (!result.Items || result.Items.length === 0) {
        return [];
    }
    
    return result.Items.map(item => ({
        tenantId: item.id?.S,
        name: item.name?.S,
        orgId: item.org_id?.S,
        userPoolId: item.userpool?.S,
        endUserClientId: item.enduserclientid?.S,
        url: item.url?.S,
        endUserSpUrl: item.endUserSpUrl?.S,
    }));
}

/**
 * Create HTTP response helper with dynamic CORS origin validation.
 *
 * When `event` is provided, the Origin header is validated against allowed
 * patterns (admin portal, SP portal, AMFA service subdomains, localhost).
 * When `event` is omitted (backward compat), falls back to '*'.
 *
 * @param {number} statusCode - HTTP status code
 * @param {Object} body - Response body object
 * @param {Object} [event] - Lambda event (optional, for dynamic CORS)
 * @returns {Object} API Gateway response
 */
export function createResponse(statusCode, body, event) {
    const headers = event
        ? getCorsHeaders(event)
        : {
            'Access-Control-Allow-Headers': 'Content-Type,Authorization,X-Api-Key,X-Requested-With,X-Tenant-Id',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'OPTIONS,GET,POST,PUT,DELETE',
            'Content-Type': 'application/json',
        };

    return {
        statusCode,
        headers,
        body: typeof body === 'string' ? body : JSON.stringify(body)
    };
}
