// tenant-utils/index.cjs
// CJS wrapper for ESM tenant-utils module
// This allows CommonJS bundled lambdas (esbuild default) to require('tenant-utils')
// Same pattern as admin-auth/index.cjs

let _module;

async function loadModule() {
  if (!_module) {
    _module = await import('./index.mjs');
  }
  return _module;
}

module.exports.getTenantInfoFromDB = async function(...args) {
  const mod = await loadModule();
  return mod.getTenantInfoFromDB(...args);
};

module.exports.getUserPoolToTenantMapping = async function(...args) {
  const mod = await loadModule();
  return mod.getUserPoolToTenantMapping(...args);
};

module.exports.getTenantId = async function(...args) {
  const mod = await loadModule();
  return mod.getTenantId(...args);
};

module.exports.getTenantUserPoolId = async function(...args) {
  const mod = await loadModule();
  return mod.getTenantUserPoolId(...args);
};

module.exports.getTenantTableNames = function(...args) {
  // Sync function — inline fallback (same logic as ESM version)
  const tenantId = args[0];
  return {
    authCodeTable: `amfa-authcode-${tenantId}`,
    sessionTable: `amfa-sessionid-${tenantId}`,
    totpTokenTable: `amfa-totptoken-${tenantId}`,
    pwdHashTable: `amfa-pwdhash-${tenantId}`,
    spInfoTable: `amfa-spinfo-${tenantId}`,
  };
};

module.exports.getAppClientSecret = async function(...args) {
  const mod = await loadModule();
  return mod.getAppClientSecret(...args);
};

module.exports.getTenantContext = async function(...args) {
  const mod = await loadModule();
  return mod.getTenantContext(...args);
};

module.exports.createTenantError = function(message, tenantId, statusCode = 500) {
  const error = new Error(message);
  error.tenantId = tenantId;
  error.statusCode = statusCode;
  return error;
};

module.exports.createResponse = function(statusCode = 200, body, tenantId = null) {
  return {
    statusCode,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': 'Content-Type,Authorization,X-Api-Key,X-Tenant-Id',
      'Access-Control-Allow-Methods': 'OPTIONS,GET,POST,PUT,DELETE',
    },
    body: typeof body === 'string' ? body : JSON.stringify(body),
  };
};
