// tenant-utils/index.mjs
/**
 * Shared utilities for multi-tenant lambda functions
 * 
 * Tenant data is stored in DynamoDB (amfa-tenanttable) and looked up at runtime.
 * The static TENANT_USER_POOLS env var is no longer used for tenant resolution.
 */

import { DynamoDBClient, GetItemCommand, QueryCommand, ScanCommand } from "@aws-sdk/client-dynamodb";
import { CognitoIdentityProviderClient, DescribeUserPoolClientCommand } from "@aws-sdk/client-cognito-identity-provider";

// Lazy-initialized DynamoDB client
let _dynamodb = null;
const getDynamoDB = () => {
  if (!_dynamodb) {
    _dynamodb = new DynamoDBClient({ region: process.env.AWS_REGION || process.env.REGION });
  }
  return _dynamodb;
};

// Lazy-initialized Cognito client
let _cognito = null;
const getCognito = () => {
  if (!_cognito) {
    _cognito = new CognitoIdentityProviderClient({ region: process.env.AWS_REGION || process.env.REGION });
  }
  return _cognito;
};

// In-memory cache for tenant info lookups
const tenantCache = new Map();
const CACHE_TTL = 5 * 60 * 1000; // 5 minutes

// Cache for userPoolId → tenantId reverse mapping (used by authorizer path)
let userPoolToTenantCache = null;
let userPoolToTenantCacheTimestamp = 0;

/**
 * Get tenant table name
 * Table naming convention: amfa-tenanttable (no account/region prefix)
 */
const getTenantTableName = () => {
  return 'amfa-tenanttable';
};

/**
 * Look up tenant info from DynamoDB with caching
 */
export const getTenantInfoFromDB = async (tenantId) => {
  const cacheKey = `tenant:${tenantId}`;
  const cached = tenantCache.get(cacheKey);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return cached.data;
  }

  const dynamodb = getDynamoDB();
  const tableName = getTenantTableName();

  try {
    // Use Query with begins_with on sk since sk format is "TENANT#<timestamp>"
    const result = await dynamodb.send(new QueryCommand({
      TableName: tableName,
      KeyConditionExpression: 'id = :id AND begins_with(sk, :sk_prefix)',
      ExpressionAttributeValues: {
        ':id': { S: `TENANT#${tenantId}` },
        ':sk_prefix': { S: 'TENANT#' }
      },
      Limit: 1
    }));

    const item = result.Items && result.Items.length > 0 ? result.Items[0] : null;
    if (!item) {
      console.warn(`[getTenantInfoFromDB] Tenant '${tenantId}' not found in ${tableName}`);
      return null;
    }

    const data = {
      userPoolId: item.userpool?.S || null,
      userPoolArn: item.userPoolArn?.S || null,
      endUserClientId: item.endUserClientId?.S || null,
      customAuthClientId: item.customAuthClientId?.S || null,
      customAuthClientSecret: item.customAuthClientSecret?.S || null,
      name: item.name?.S || null,
    };

    tenantCache.set(cacheKey, { data, timestamp: Date.now() });
    return data;
  } catch (error) {
    console.error(`[getTenantInfoFromDB] Error looking up tenant '${tenantId}':`, error);
    return null;
  }
};

/**
 * Build a reverse mapping of userPoolId → tenantId by scanning the tenant table.
 * Cached for CACHE_TTL.
 */
export const getUserPoolToTenantMapping = async () => {
  if (userPoolToTenantCache && Date.now() - userPoolToTenantCacheTimestamp < CACHE_TTL) {
    return userPoolToTenantCache;
  }

  const dynamodb = getDynamoDB();
  const tableName = getTenantTableName();
  const mapping = {};

  try {
    const result = await dynamodb.send(new ScanCommand({
      TableName: tableName,
      FilterExpression: 'begins_with(id, :prefix)',
      ExpressionAttributeValues: {
        ':prefix': { S: 'TENANT#' }
      },
      ProjectionExpression: 'id, userpool, endUserClientId'
    }));

    for (const item of (result.Items || [])) {
      const tenantId = item.id?.S?.replace('TENANT#', '');
      const userPoolId = item.userpool?.S;
      if (tenantId && userPoolId) {
        mapping[userPoolId] = {
          tenantId,
          endUserClientId: item.endUserClientId?.S || null,
        };
      }
    }

    userPoolToTenantCache = mapping;
    userPoolToTenantCacheTimestamp = Date.now();
    console.log(`[getUserPoolToTenantMapping] Loaded ${Object.keys(mapping).length} tenant mappings`);
  } catch (error) {
    console.error('[getUserPoolToTenantMapping] Error scanning tenant table:', error);
  }

  return mapping;
};

/**
 * Extract tenant ID from API Gateway request context
 * Supports both authenticated (via JWT issuer) and unauthenticated (via origin URL) requests
 */
export const getTenantId = async (event) => {
  // Method 1: Try to get from authorizer context (authenticated requests)
  if (event.requestContext?.authorizer?.tenantId) {
    // Custom authorizer already resolved tenant
    console.log(`[getTenantId] Found tenant from authorizer context: ${event.requestContext.authorizer.tenantId}`);
    return event.requestContext.authorizer.tenantId;
  }

  if (event.requestContext?.authorizer?.issuer) {
    const issuer = event.requestContext.authorizer.issuer;
    
    // Extract user pool ID from issuer URL
    const userPoolIdMatch = issuer.match(/\/([^\/]+)$/);
    if (userPoolIdMatch) {
      const userPoolId = userPoolIdMatch[1];
      const mapping = await getUserPoolToTenantMapping();
      const tenantInfo = mapping[userPoolId];
      if (tenantInfo) {
        console.log(`[getTenantId] Found tenant from JWT issuer: ${tenantInfo.tenantId}`);
        return tenantInfo.tenantId;
      }
    }
  }

  // Method 2: Fallback for unauthenticated requests - extract from origin URL
  const origin = event.headers?.origin || event.headers?.Origin;
  const rootDomain = process.env.ROOT_DOMAIN_NAME;

  if (origin && rootDomain) {
    try {
      const originUrl = new URL(origin);
      const originHost = originUrl.hostname;
      
      if (originHost.endsWith(`.${rootDomain}`)) {
        let tenantId = originHost.replace(`.${rootDomain}`, '');
        
        // Strip known subdomain prefixes (e.g., "testtenantb.idapersona" → "testtenantb")
        const knownSubdomains = ['idapersona', 'login'];
        for (const sub of knownSubdomains) {
          if (tenantId.endsWith(`.${sub}`)) {
            tenantId = tenantId.replace(`.${sub}`, '');
            break;
          }
        }
        
        if (tenantId && tenantId.length > 0) {
          console.log(`[getTenantId] Found tenant from origin URL: ${tenantId}`);
          return tenantId;
        } else {
          console.warn(`[getTenantId] Empty tenant ID extracted from origin URL`);
        }
      } else {
        console.warn(`[getTenantId] Origin host '${originHost}' does not match root domain '${rootDomain}'`);
      }
    } catch (error) {
      console.error('[getTenantId] Error parsing origin URL:', error);
    }
  } else {
    console.warn(`[getTenantId] Missing origin (${origin}) or ROOT_DOMAIN_NAME (${rootDomain})`);
  }

  throw new Error("Tenant ID not found in request context or origin URL");
};

/**
 * Get user pool ID for a specific tenant
 * Looks up from DynamoDB (with cache) since tenants are provisioned at runtime
 */
export const getTenantUserPoolId = async (tenantId) => {
  const tenantInfo = await getTenantInfoFromDB(tenantId);
  
  if (tenantInfo?.userPoolId) {
    return tenantInfo.userPoolId;
  }

  console.warn(`[getTenantUserPoolId] User pool not found for tenant: ${tenantId}`);
  return null;
};

/**
 * Construct tenant-specific table names dynamically
 * Naming convention:
 *   Shared: amfa-configtable, amfa-tenanttable
 *   Per-tenant: amfa-authcode-{tenantId}, amfa-sessionid-{tenantId}, etc.
 */
export const getTenantTableNames = (tenantId) => {
  return {
    // Per-tenant tables
    authCodeTable: `amfa-authcode-${tenantId}`,
    sessionIdTable: `amfa-sessionid-${tenantId}`,
    totpTokenTable: `amfa-totptoken-${tenantId}`,
    pwdHashTable: `amfa-pwdhash-${tenantId}`,
    // Shared tables (same for all tenants)
    configTable: `amfa-configtable`,
    tenantTable: `amfa-tenanttable`,
    amfaConfigTable: `amfa-configtable`,
  };
};

/**
 * Get the Cognito app client secret for a given userPoolId + clientId.
 * Cached per clientId for CACHE_TTL.
 */
const clientSecretCache = new Map();

export const getAppClientSecret = async (userPoolId, clientId) => {
  if (!userPoolId || !clientId) {
    console.warn(`[getAppClientSecret] Missing userPoolId (${userPoolId}) or clientId (${clientId})`);
    return null;
  }

  const cacheKey = `clientSecret:${clientId}`;
  const cached = clientSecretCache.get(cacheKey);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return cached.data;
  }

  try {
    const cognito = getCognito();
    const result = await cognito.send(new DescribeUserPoolClientCommand({
      UserPoolId: userPoolId,
      ClientId: clientId,
    }));

    const secret = result.UserPoolClient?.ClientSecret || null;
    clientSecretCache.set(cacheKey, { data: secret, timestamp: Date.now() });
    console.log(`[getAppClientSecret] Retrieved client secret for clientId: ${clientId}`);
    return secret;
  } catch (error) {
    console.error(`[getAppClientSecret] Error fetching client secret for clientId '${clientId}':`, error);
    return null;
  }
};

/**
 * Get all tenant context needed for lambda functions
 * Returns tenantId, tableNames, userPoolId, domainName, appClientId, appSecret, magicString
 */
export const getTenantContext = async (event) => {
  const tenantId = await getTenantId(event);
  const tableNames = getTenantTableNames(tenantId);
  const tenantInfo = await getTenantInfoFromDB(tenantId);

  const userPoolId = tenantInfo?.userPoolId || null;

  // Primary: read customAuthClientId/Secret from DDB (stored during provisioning)
  let appClientId = tenantInfo?.customAuthClientId || null;
  let appSecret = tenantInfo?.customAuthClientSecret || null;

  // Fallback for existing tenants provisioned before customAuthClient was stored in DDB:
  // Use DescribeUserPoolClient to fetch the secret at runtime
  if (appClientId && !appSecret) {
    appSecret = await getAppClientSecret(userPoolId, appClientId);
  }

  if (!appClientId) {
    console.warn(`[getTenantContext] customAuthClientId not found in DDB for tenant '${tenantId}'. Tenant may need re-provisioning.`);
  }

  return {
    tenantId,
    tableNames,
    userPoolId,
    domainName: process.env.ROOT_DOMAIN_NAME || '',
    appClientId,
    appSecret,
    magicString: process.env.MAGIC_STRING || 'youguesswhat',
  };
};

/**
 * Enhanced error handling with tenant context
 */
export const createTenantError = (message, tenantId, statusCode = 500) => {
  const error = new Error(`[Tenant: ${tenantId}] ${message}`);
  error.statusCode = statusCode;
  error.tenantId = tenantId;
  return error;
};

/**
 * Standard response helper with CORS headers
 */
export const createResponse = (statusCode = 200, body, tenantId = null) => {
  const headers = {
    "Access-Control-Allow-Headers":
      "Content-Type,Authorization,X-Api-Key,X-Requested-With,X-Tenant-Id",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "OPTIONS,GET,POST,PUT,DELETE",
  };

  if (tenantId) {
    headers["X-Tenant-Id"] = tenantId;
  }

  console.log(`[Tenant: ${tenantId || "unknown"}] Response:`, {
    statusCode,
    headers,
    body: typeof body === "string" ? body : JSON.stringify(body),
  });

  return {
    statusCode,
    headers,
    body: typeof body === "string" ? body : JSON.stringify(body),
  };
};
